oeps/oep-0017-feature-toggles.rst

```diff
@@ -0,0 +1,809 @@
+  a fact to be buggy, untested, incomplete and quite possibly incompatible with 
+  your live data. Your if statements and configuration settings are themselves 
+  code which is subject to bugs – and furthermore can only be tested in production.
+  ...
+  Your features may not be as isolated from each other as you thought they were, 
+  and you may end up deploying bugs to your production environment.”
+
+.. _James McKay puts it: https://jamesmckay.net/2011/07/why-does-martin-fowler-not-understand-feature-branches/
+
+Testing Best Practices
+~~~~~~~~~~~~~~~~~~~~~~
+
+Given that, here are best practices for testing a Feature Toggle:
+
+* Tests should run with whatever state is in production.
+
+* Tests should run in both on and off Toggle states unless they are guaranteed
+  to not be enabled in production.
+
+  * Acceptance or end-to-end tests for Toggles that gate user-facing changes
+    should also be run in both on and off Toggle states.
+
+  * Browser-based automation (e.g., Selenium_) tests should be able to:
+
+    * determine the state of a Toggle by calling a REST API (e.g., `wafflejs API`_
+      using WaffleJS_).
+
+    * override a Toggle value by passing in the desired value in a request parameter
+      (e.g., `Overriding Flags`_).
+
+* Test environments, such as Devstack_ and central Staging should allow incoming
+  requests to override Toggles (e.g., by setting WAFFLE_OVERRIDE_).
+
+.. _Selenium: http://www.seleniumhq.org/
+.. _wafflejs API: https://github.com/jsocol/django-waffle/blob/f4b10e8c46cf5305fbaa082b169f221fdab72162/waffle/urls.py#L8
+.. _WaffleJS: https://waffle.readthedocs.io/en/stable/usage/javascript.html
+.. _Overriding Flags: https://waffle.readthedocs.io/en/v0.9/usage.html#overriding-flags
+.. _Devstack: https://github.com/edx/devstack
+.. _WAFFLE_OVERRIDE: https://waffle.readthedocs.io/en/v0.9/usage.html#waffle-override
+
+Test Plans for Toggle Use Cases
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following diagram summarizes test plans for the various toggle use cases while
+taking best practices into consideration.
+
+.. image:: oep-0017/test_plan_for_use_cases.png
+   :alt: A diagram that shows the test plans for the various toggle use cases.
+
+Removal
+-------
+
+As mentioned previously, feature toggles inherently bring along code complexity.
+In order to manage the "toggle debt", we need to keep their inventory at a
+minimum. The framework's Removability and Report features make it possible
+to do so. But it must be accompanied by a proactive process of actually removing
+the toggles and their branches within the code.
+
+In addition to using the Report as a central tool of overseeing the toggles,
+individual teams should create tickets in their backlogs for removing toggles
+according to their intended expiration dates.
+
+Rationale
+=========
+
+Although feature toggles have been in use from the very early stages of development
+on the platform, the `Feature Flags and Settings on edx-platform`_ wiki was one
```

Comment:
Regarding documentation location:

- This OEP is the place for best practices on feature toggles.
- We will have an OEP on best practices on tests.
- There will be some overlap and cross-references between the Testing OEP and this one.
- Implementation details about the waffle_utils API should live within the waffle_utils folder.

Regarding the additional details, as I said above, it would be best if we remove the option to have a "default" value for toggles.  I believe it will add more confusion and bugs, which outweigh its convenience.

Reply:
I'm fine with removing the "default" value, but only if we have an alternate suggestion that satisfies the requirements I stated above.  Maybe it is through a migration that adds the flag?  I can't remember why we went down the path we did, but in all cases, I think we should help people understand the path and what to do.

Reply:
Sorry if I missed it @nasthagiri, but I didn't see an update related to the conversation about when and how people need to deal with this situation. We had discussed a preference of migrations over the default functionality (which may need to be deprecated). We also discussed @jmbowman's idea of managing these, but I don't remember the details.

Reply:
I attached @efischer19's notes, which includes possibilities that were discussed.  However, we decided to wait until the next time this comes up before we recommend a best practice for it - as it needs to be further fleshed out - if it comes up again.
From my own multiple experiences with using waffle, we haven't had the need for specifying a default or for running a migration.  The need may be there for a particular Use Case that I haven't run into myself.